### PR TITLE
Fetch visible rows only

### PR DESCRIPTION
**Pull request by PatrickTasse**
Fetch visible rows only

Add row provider interface to fetch the full list of row ids.

Add selected row index to TimeGraphRowController to allow for the case
where the selected row model is not yet available. Allow for selected
row model to be undefined.

When the list of row ids changes, delete row component for removed rows
and update the y-position of existing rows. Update the selected row
index. Create placeholder rows for new rows until their row model is
fetched.

Allow for placeholder rows to be created and displayed while the row
model is being fetched, by allowing constructor with undefined row
model.

Make each TimeGraphRowComponent have its own providedModel (range and
resolution) object so that each row can determine independently if it
needs to fetch new data when the view range or resolution changes.

Compute the list of visible row ids (with a few extra buffered rows) and
filter out those that already have the correct model. Fetch the row
model of only the necessary rows.

When the row models are received, destroy only the row components to be
updated and recreate them. The other row components are left untouched.

Move states and annotations from the TimeGraphChartLayer to be children
of their respective TimeGraphRowComponent instead.

If a newly created row component is at the selected row index, select
it.

In updateScaleAndPosition(), iterate through existing row components
instead of row models. Handle placeholder rows that have no states or
annotations.

Allow row style provider to provide default style for placeholder rows
when row model is undefined.

Call the debounced maybeFetchNewData() method when the vertical offset
is changed to fetch row model of rows that become visible.

Remove getRowModels() method as all row models are not always available.
Add getRowModel(index) method.

Replace getElementById() with getStateById(). Iterate through the row
components to find the state.

Move arrow up/down navigation from TimeGraphChartCursors layer to
TimeGraphChart layer. Ensure the newly selected row is visible.

Make options optional in update of annotation.

Signed-off-by: Patrick Tasse <patrick.tasse@gmail.com>

Fix use of TimeGraphArrow sourceId and destinationId

The sourceId and destinationId should be the row id of source row and
destination row. But it is used as the index of the source and
destination rows in the full list of rows.

Change to use the sourceId and destinationId properly. This requires
that the TimeGraphChartArrows layer receive the full list of row ids.

Enable unit tests in build pipeline.

Signed-off-by: Patrick Tasse <patrick.tasse@gmail.com>

### CODE REVIEW

**Review by bhufmann — CHANGES_REQUESTED** [review]
Thanks for the improvement. It works quite well. However, I have some initial findings that need to be fixed.

File: timeline-chart/src/layer/time-graph-chart.ts [hunk 1]
```diff
@@ -547,13 +593,7 @@ export class TimeGraphChart extends TimeGraphChartLayer {
                 const el = this.createNewState(stateModel, rowComponent);
                 if (el) {
                     this.addElementInteractions(el);
```
[the remark this answers is not in the document]
Comment (PatrickTasse) [reply timeline-chart/src/layer/time-graph-chart.ts:595]: I think this one should stay addElementInteractions, it's also used for annotations. And maybe eventually for arrows?

File: timeline-chart/src/layer/time-graph-chart.ts [hunk 2]
```diff
@@ -791,4 +830,60 @@ export class TimeGraphChart extends TimeGraphChartLayer {
         }
         this.setNavigationFlag(false);
     }
+
+    protected navigateDown() {
+        if (this.rowIds.length > 0) {
```
[the remark this answers is not in the document]
Comment (PatrickTasse) [reply timeline-chart/src/layer/time-graph-chart.ts:835]: I added a description at the declaration.

**Review by bhufmann — CHANGES_REQUESTED** [review]
I have one question left. The commit messages doesn't mention about the code for arrow-up/down navigation being moved to another class. Please add it.

**Comment by bhufmann** [comment]
> I have one question left. The commit messages doesn't mention about the code for arrow-up/down navigation being moved to another class. Please add it.

I missed it the commit message. It's already there

File: timeline-chart/src/layer/time-graph-chart.ts [hunk 3]
```diff
@@ -19,10 +19,11 @@ export interface TimeGraphMouseInteractions {
 }
 
 export interface TimeGraphChartProviders {
-    dataProvider: (range: TimelineChart.TimeGraphRange, resolution: number) => Promise<{ rows: TimelineChart.TimeGraphRowModel[], range: TimelineChart.TimeGraphRange, resolution: number }> | { rows: TimelineChart.TimeGraphRowModel[], range: TimelineChart.TimeGraphRange, resolution: number } | undefined
+    rowProvider: () => { rowIds: number[] }
```
[the remark this answers is not in the document]
Comment (PatrickTasse) [reply timeline-chart/src/layer/time-graph-chart.ts:22]: The implementation of rowProvider in theia-trace-extension returns immediately from an already-fetched tree model, it does not need to communicate with the trace-server.

Further, making this async would require adding protection against out-of-order responses.

For these reasons I think it's best not to change to a Promise.

[the remark this answers is not in the document]
Comment (bhufmann) [reply timeline-chart/src/layer/time-graph-chart.ts:22]: Ok, thanks for the explanation.

**Review by bhufmann — APPROVED** [review]
Looks good to me. Works well with the theia-trace-extension integration and brings performance improvements for it.

**Review by Rodrigoplp-work — APPROVED** [review]

**Pull request merged**